### PR TITLE
feat(SPV-846): refactor how the errors are returned

### PR DESCRIPTION
**Pull request by pawellewandowski98**
# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/go-paymail/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/go-paymail/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

# Description
- Define all errors which are returned from endpoints
- Refactor the way of returning errors

### CODE REVIEW

File: errors/definitions.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	ErrLockTimeAndSequence = models.SPVError{Message: "nLocktime is set and nSequence is not max, therefore this could be a non-final tx which is not currently supported", StatusCode: 417, Code: "error-spv-locktime-sequence"}
+
+	// ErrOutputValueTooHigh is when the satoshis output is too high on a transaction
+	ErrOutputValueTooHigh = models.SPVError{Message: "invalid input and output sum, outputs can not be larger than inputs", StatusCode: 417, Code: "error-spv-output-value-too-high"}
+
+	// ErrBUMPAncestorNotPresent is when the input mined ancestor is not present in BUMPs
+	ErrBUMPAncestorNotPresent = models.SPVError{Message: "invalid BUMP - input mined ancestor is not present in BUMPs", StatusCode: 417, Code: "error-spv-bump-ancestor-not-present"}
+
+	// ErrBUMPCouldNotFindMindParent is when the mined parent for input could not be found
+	ErrBUMPCouldNotFindMindParent = models.SPVError{Message: "invalid BUMP - cannot find mined parent for input", StatusCode: 417, Code: "error-spv-bump-ancestor-not-present"}
```
Comment (ac4ch) [inline errors/definitions.go:131]: typo ? `ErrBUMPCouldNotFindMinedParent`

File: errors/definitions.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	ErrServiceProviderNil = models.SPVError{Message: "service provider is nil", StatusCode: 500, Code: "error-configuration-service-provider-nil"}
+)
+
+// CAPABILITY ERRORS
+var (
+	//ErrPrefixOrDomainMissing is when the prefix or domain is missing
+	ErrPrefixOrDomainMissing = models.SPVError{Message: "prefix or domain is missing", StatusCode: 400, Code: "error-capabilities-prefix-or-domain-missing"}
+
+	//ErrDomainUnknown is when the domain is not in the list of allowed domains
+	ErrDomainUnknown = models.SPVError{Message: "prefix or domain is missing", StatusCode: 400, Code: "error-capabilities-prefix-or-domain-missing"}
```
Comment (ac4ch) [inline errors/definitions.go:34]: this error has same message and code as `ErrPrefixOrDomainMissing`

File: errors/definitions.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	ErrOutputValueTooHigh = models.SPVError{Message: "invalid input and output sum, outputs can not be larger than inputs", StatusCode: 417, Code: "error-spv-output-value-too-high"}
+
+	// ErrBUMPAncestorNotPresent is when the input mined ancestor is not present in BUMPs
+	ErrBUMPAncestorNotPresent = models.SPVError{Message: "invalid BUMP - input mined ancestor is not present in BUMPs", StatusCode: 417, Code: "error-spv-bump-ancestor-not-present"}
+
+	// ErrBUMPCouldNotFindMindParent is when the mined parent for input could not be found
+	ErrBUMPCouldNotFindMindParent = models.SPVError{Message: "invalid BUMP - cannot find mined parent for input", StatusCode: 417, Code: "error-spv-bump-ancestor-not-present"}
+
+	// ErrNoMatchingTransactionsForInput is when no matching transaction for input can be found
+	ErrNoMatchingTransactionsForInput = models.SPVError{Message: "invalid parent transactions, no matching transactions for input", StatusCode: 417, Code: "error-spv-bump-ancestor-not-present"}
```
Comment (ac4ch) [inline errors/definitions.go:134]: not sure but i would use different code not same as 2 above

File: errors/http_response.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+// ErrorResponse is a standard way to return errors to the client
+func ErrorResponse(c *gin.Context, err error) {
+	response, statusCode := getError(err)
+	c.JSON(statusCode, response)
+}
+
+func getError(err error) (models.ResponseError, int) {
+	var errDetails models.SPVError
```
Comment (ac4ch) [inline errors/http_response.go:16]: check for nil here 
```go
if err == nil {
		return models.ResponseError{Code: models.UnknownErrorCode, Message: "No error information available"}, 500
	}
```

File: go.mod [hunk 5]
```diff
@@ -1,8 +1,9 @@
 module github.com/bitcoin-sv/go-paymail
 
-go 1.21.5
+go 1.22.4
```
Comment (ac4ch) [inline go.mod:3]: intentional ? if yes u can resolve this straightaway :)

**Review by ac4ch — COMMENTED** [review]
only few comments :)

File: go.mod (hunk 5 above)
Comment (chris-4chain) [reply go.mod:3]: I would prefer if we bump this golang version in separate task.
Or you're completely sure that it won't break anything?

File: go.mod [hunk 6]
```diff
@@ -1,8 +1,9 @@
...
 
 require (
+	github.com/bitcoin-sv/spv-wallet/models v1.0.0-beta.14
```
Comment (chris-4chain) [inline go.mod:6]: As I recall our conversation, we wanted not to make `go-paymail` dependant to `spv-wallet/models`.

Maybe in this repo we could introduce the same `SPVError`, like `PaymailError` and in `spv-wallet/engine/spverrors/getError` handle also this new struct 🤔 ?

File: errors/definitions.go (hunk 2 above)
Comment (pawellewandowski98) [reply errors/definitions.go]: Fixed

File: errors/definitions.go (hunk 3 above)
Comment (pawellewandowski98) [reply errors/definitions.go:140]: Fixed

File: errors/http_response.go (hunk 4 above)
Comment (pawellewandowski98) [reply errors/http_response.go:20]: Added

File: errors/definitions.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
+package errors
+
+import (
+	"github.com/bitcoin-sv/spv-wallet/models"
```
Comment (dorzepowski) [inline errors/definitions.go:4]: Unfortunately I must say that I don't like the idea to add a dependency to models here (neither the idea to create separate package for error model and have it in dependencies 🙅 ).

My suggestion would be to choose between:
a. change models.SPVError to be some interface (add some interface that SPVError would implement). Then handle errors based on that interface in spv-wallet. And here just make the same methods as this interface has, so it should fit it out of the box.
b. or add here similar struct and handle it in similar way as SPVErrror in spv-wallet
c. or translate go-paymail errors into SPVError in spv-wallet

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: go.mod (hunk 5 above)
Comment (chris-4chain) [reply go.mod:3]: Yea - and it has to be bumped to use new models.

Comment (pawellewandowski98) [reply go.mod]: Done, I went back to the old version

File: errors/definitions.go (hunk 7 above)
Comment (pawellewandowski98) [reply errors/definitions.go]: Done, I chose option b. Added models to errors directory and removed models package

Comment (pawellewandowski98) [reply errors/definitions.go]: After tests code as refactored to option a

**Review by dorzepowski — APPROVED** [review]

**Review by chris-4chain — APPROVED** [review]

**Pull request merged**